[pytorch] improve code analyzer to dump ops called from c++ functions

ljk53 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #36607 [pytorch] handle pybind11 style registration API with code analyzer
* **#35941 [pytorch] improve code analyzer to dump ops called from c++ functions**

Summary:
The key step of mobile custom build is to find out ops used by specific
model, with which it can produce a tailored build of optimal size.

However, ops can not only be called from TorchScript model but can also
be called from C++ code directly, e.g.: via torch::jit:: APIs. With
static dispatch, ops called this way will be statically linked into client
code. With dynamic dispatch, we need obtain & keep these ops explicitly.

This PR improves static code analyzer to dump ops that are called from
visible c++ symbols matching specific regex. This provides a mechanism
to solve the custom build problem with dynamic dispatch.

It starts with dumping ops that are callable from functions in torch::jit
namespace and include them in custom build with dynamic dispatch. We can
extend it to analyze custom code / to refine the set of JIT APIs that
are relevant, and etc. This is just a preliminary version. We need
improve its usability for more general purpose.

Differential Revision: [D20835166](https://our.internmc.facebook.com/intern/diff/D20835166)

ljk53 committed:
[pytorch] improve code analyzer to dump ops called from c++ functions

Summary:
The key step of mobile custom build is to find out ops used by specific
model, with which it can produce a tailored build of optimal size.

However, ops can not only be called from TorchScript model but can also
be called from C++ code directly, e.g.: via torch::jit:: APIs. With
static dispatch, ops called this way will be statically linked into client
code. With dynamic dispatch, we need obtain & keep these ops explicitly.

This PR improves static code analyzer to dump ops that are called from
visible c++ symbols matching specific regex. This provides a mechanism
to solve the custom build problem with dynamic dispatch.

It starts with dumping ops that are callable from functions in torch::jit
namespace and include them in custom build with dynamic dispatch. We can
extend it to analyze custom code / to refine the set of JIT APIs that
are relevant, and etc. This is just a preliminary version. We need
improve its usability for more general purpose.

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch] improve code analyzer to dump ops called from c++ functions"


Summary:
The key step of mobile custom build is to find out ops used by specific
model, with which it can produce a tailored build of optimal size.

However, ops can not only be called from TorchScript model but can also
be called from C++ code directly, e.g.: via torch::jit:: APIs. With
static dispatch, ops called this way will be statically linked into client
code. With dynamic dispatch, we need obtain & keep these ops explicitly.

This PR improves static code analyzer to dump ops that are called from
visible c++ symbols matching specific regex. This provides a mechanism
to solve the custom build problem with dynamic dispatch.

It starts with dumping ops that are callable from functions in torch::jit
namespace and include them in custom build with dynamic dispatch. We can
extend it to analyze custom code / to refine the set of JIT APIs that
are relevant, and etc. This is just a preliminary version. We need
improve its usability for more general purpose.

[ghstack-poisoned]

ezyang commented on tools/code_analyzer/run_analyzer.sh:
```diff
@@ -12,6 +12,7 @@ echo "Analyze: ${INPUT}"
   -op_schema_pattern="^(_aten|_prim|aten|quantized|profiler|_test)::[^ ]+" \
   -op_register_pattern="c10::RegisterOperators::(op|checkSchemaAndRegisterOp_)|c10::Module::(def|impl)" \
   -op_invoke_pattern="c10::Dispatcher::findSchema|callOp" \
+  -root_symbol_pattern="torch::jit::[^(]" \
```
Why is this `torch::jit`?

ezyang commented on tools/code_analyzer/op_dependency.cpp:
```diff
@@ -126,6 +126,15 @@ cl::opt<RegexOpt, true, cl::parser<std::string>> OpInvocationPattern(
+    cl::desc("Regular expression used to identify root symbols. It will insert "
+             "an entry to the output graph with key = `__ROOT__` and value = "
+             "set of ops reachable from root symbols, if the pattern is set. "
+             "Example: -root_symbol_pattern 'torch::jit'"),
```
I think it's still not altogether clear to me what I, as a user, should put here. At a guess, it sounds like a C++ user of the analysis tool should specify a root symbol to be some "reasonably initial" symbol, from which we take the transitive (static?) closure to look for all operator calls. So as a C++ user should I put "main" as the root symbol? But the example here suggests maybe I should be namespacing my C++ API calls in such a way that I can just match for that namespace? What's the idea?

ljk53 replied:
I added answers to these questions to the code as comment:
```
// Why not dump ops that are reachable from any visible C++ symbols? Why
// limit it to a subset of root symbols?
// Because op registration callsites in static initializer are visible root
// symbols, too. It will dump ALL the registered ops without any filtering.
//
// Can we use some fixed entry point like `main()`?
// The target to be analyzed can be DSO that doesn't have a `main()`.
```

"torch::jit" is a simple pattern I use to cover most visible torch APIs that could be called externally (might not actually be called in the binary that is being analyzed).

We can refine the pattern later on with more accurate set of symbols. It's only exposed to be called by the checked-in shell script - I'm not sure whether we want to expose this flag to external usage, yet.

Won't we need multiple root symbol patterns in the general case? Yes, I guess we can concatenate the symbols into one RE with "|"?

bhosmer reviewed (approved):
This LGTM as a near-term enhancement to keep the current analysis working, per offline discussions. One question inline but not blocking. Though before landing I'd be interested to hear if Ed brings any recommendations back from his tire-kicking session :)

ezyang commented on tools/code_analyzer/op_dependency.cpp:
```diff
@@ -201,32 +210,42 @@ class OpDependency : public ModulePass {
     // dependency graph and to find out instructions that might register or
     // invoke operators, respectively.
```
this comment should get updated

ezyang reviewed (approved):
Code changes look good, most of the request is for clearer documentation story

ljk53 committed:
Update on "[pytorch] improve code analyzer to dump ops called from c++ functions"


Summary:
The key step of mobile custom build is to find out ops used by specific
model, with which it can produce a tailored build of optimal size.

However, ops can not only be called from TorchScript model but can also
be called from C++ code directly, e.g.: via torch::jit:: APIs. With
static dispatch, ops called this way will be statically linked into client
code. With dynamic dispatch, we need obtain & keep these ops explicitly.

This PR improves static code analyzer to dump ops that are called from
visible c++ symbols matching specific regex. This provides a mechanism
to solve the custom build problem with dynamic dispatch.

It starts with dumping ops that are callable from functions in torch::jit
namespace and include them in custom build with dynamic dispatch. We can
extend it to analyze custom code / to refine the set of JIT APIs that
are relevant, and etc. This is just a preliminary version. We need
improve its usability for more general purpose.

Differential Revision: [D20835166](https://our.internmc.facebook.com/intern/diff/D20835166)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch] improve code analyzer to dump ops called from c++ functions"


Summary:
The key step of mobile custom build is to find out ops used by specific
model, with which it can produce a tailored build of optimal size.

However, ops can not only be called from TorchScript model but can also
be called from C++ code directly, e.g.: via torch::jit:: APIs. With
static dispatch, ops called this way will be statically linked into client
code. With dynamic dispatch, we need obtain & keep these ops explicitly.

This PR improves static code analyzer to dump ops that are called from
visible c++ symbols matching specific regex. This provides a mechanism
to solve the custom build problem with dynamic dispatch.

It starts with dumping ops that are callable from functions in torch::jit
namespace and include them in custom build with dynamic dispatch. We can
extend it to analyze custom code / to refine the set of JIT APIs that
are relevant, and etc. This is just a preliminary version. We need
improve its usability for more general purpose.

Differential Revision: [D20835166](https://our.internmc.facebook.com/intern/diff/D20835166)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch] improve code analyzer to dump ops called from c++ functions"


Summary:
The key step of mobile custom build is to find out ops used by specific
model, with which it can produce a tailored build of optimal size.

However, ops can not only be called from TorchScript model but can also
be called from C++ code directly, e.g.: via torch::jit:: APIs. With
static dispatch, ops called this way will be statically linked into client
code. With dynamic dispatch, we need obtain & keep these ops explicitly.

This PR improves static code analyzer to dump ops that are called from
visible c++ symbols matching specific regex. This provides a mechanism
to solve the custom build problem with dynamic dispatch.

It starts with dumping ops that are callable from functions in torch::jit
namespace and include them in custom build with dynamic dispatch. We can
extend it to analyze custom code / to refine the set of JIT APIs that
are relevant, and etc. This is just a preliminary version. We need
improve its usability for more general purpose.

Differential Revision: [D20835166](https://our.internmc.facebook.com/intern/diff/D20835166)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch] improve code analyzer to dump ops called from c++ functions"


Summary:
The key step of mobile custom build is to find out ops used by specific
model, with which it can produce a tailored build of optimal size.

However, ops can not only be called from TorchScript model but can also
be called from C++ code directly, e.g.: via torch::jit:: APIs. With
static dispatch, ops called this way will be statically linked into client
code. With dynamic dispatch, we need obtain & keep these ops explicitly.

This PR improves static code analyzer to dump ops that are called from
visible c++ symbols matching specific regex. This provides a mechanism
to solve the custom build problem with dynamic dispatch.

It starts with dumping ops that are callable from functions in torch::jit
namespace and include them in custom build with dynamic dispatch. We can
extend it to analyze custom code / to refine the set of JIT APIs that
are relevant, and etc. This is just a preliminary version. We need
improve its usability for more general purpose.

Differential Revision: [D20835166](https://our.internmc.facebook.com/intern/diff/D20835166)

[ghstack-poisoned]
